### ISSUE TITLE
BulkCopy and KeepIdentity

### ISSUE THREAD

**Issue by YXMagnuMXY** [linq2db/linq2db#1036]
Hi,
I am using Linq2db and i want to do a BulkCopy instead of inserting in the DB each instance from a list.
So, the list I want to insert is "ListToCopy"(Data taken from other DB and has ID's).
Now i want to copy this list to Other DB. When i use BulkCopy, is inserting all the instances from list with new ID's... :( Even KeepIdentity setted to true is not helping...

```c#
Database.MyClass.BulkCopy(new BulkCopyOptions { KeepIdentity = true }, ListToCopy);
```

"ListToCopy" has ID's but are inserted in Other DB with new ID's.
Is there a posibility to copy a list of data into a DB including the actual ID's?
Please help...

PS: PostgreSQL data provider for both databases.

**Comment by Franki1986** [comment]
You could try the [Merge Api](https://github.com/linq2db/linq2db/wiki/Merge-API)

**Comment by sdanyliv** [comment]
Check that column do not have `SkipOnInsert`.
Also you can create new type that do not have these attributes and insert using this technique:
```cs
db.BulkCopy(db.GetTable<MyNewClass>().TableName("MyTable"), items)
```
Also, possible, you need to execute additional statement, i'm not familiar with PostgreSQL
```cs
db.Execute("SET IDENTITY_INSERT MyTable ON")
```

**Comment by YXMagnuMXY** [comment]
Also - If i'm using the lower code i can insert each instance with ID
'foreach (var activeTest in ActiveTestsCopyList)
                {
                    tempDb.ActiveTests
                        .Value(p => p.Id, activeTest.Id)
                        .Value(p => p.PositionNumber, activeTest.PositionNumber)
                        .Value(p => p.Status, activeTest.Status)
                        .Value(p => p.ClosedDate, activeTest.ClosedDate)
                        .Value(p => p.AssignedToId, activeTest.AssignedToId)
                        .Value(p => p.ControlTypeId, activeTest.ControlTypeId)
                        .Value(p => p.TestId, activeTest.TestId)
                        .Value(p => p.ZoneId, activeTest.ZoneId)
                        .Value(p => p.Order, activeTest.Order)
                        .Insert();
                   tempCount++;
                }'
But that means i am creating many Inserts commands. But why i should do so, if i can use BulkCopy with KeepIdentity? 💃 
But BulkCopy is not copying the identity.
I will try your first step.
Also what did you mean by SkipOnInsert?

**Comment by sdanyliv** [comment]
> Also what did you mean by SkipOnInsert?

ColumnAttribute may have property set to `true`, it also maybe IsIdentity is set to true.
Creating additional class should help, but remove `IsIdentity` from attributes.

**Comment by YXMagnuMXY** [comment]
Yes, that's working. 
BUT only in case if we want to copy only one or two tables.
In my case i'm sorting data from 7 tables, and after that i'm inserting.
That means that i should create new classes without `IsIdentity` atribute near the `ID` column.
This is not so good...also this will be only used when i need to copy data from One DB to Other DB. If tomorrow i will need to copy new table, i will need to create a new class to perform correct `BulkCopy,` right?
Also, Why then the `KeepIdentity` in `BulkCopy` is doing?
Without `KeepIdentity = false` in `BulkCopy` will the `Query` in DB should look like:
`INSERT INTO MyTable ( Column1, Column2 ) VALUES
( Value1, Value2 ), ( Value1, Value2 )`
and with KeepIdentity = true should look like:
`INSERT INTO MyTable ( ID, Column1, Column2 ) VALUES
( 71, Value1, Value2 ), ( 78, Value1, Value2 )`

Right? )) Is just like adding one more column and value in the `Query`.

**Comment by sdanyliv** [comment]
I don't know why, but BulkCopy in PosgreSql implementation ignores this parameter.

**Comment by YXMagnuMXY** [comment]
Is there a probability of fixing this to the next version?

**Comment by sdanyliv** [comment]
@YXMagnuMXY, looks like it was long standing bug. I have created PR, after review it will be merged.

**Comment by YXMagnuMXY** [comment]
@sdanyliv Thank you!

**Comment by YXMagnuMXY** [comment]
@sdanyliv PS: Also "CheckConstraints" is not working, and maybe the others also...

**Comment by sdanyliv** [comment]
They are introduced for SQL Server. If you have idea how to mimique them for PostgreSQL - help is appreciated.
Linq2db has long evolution path, so many databases and their providers may be improved since last implementation.

**Comment by YXMagnuMXY** [comment]
Hmm... I think KeepIdentity is a restriction, and excludes the Identity column from the table.
Like i showed before, the first query is converted from a linq2db expression to SQL expression that can be understand by Postgres Server.
`INSERT INTO MyTable ( Column1, Column2 ) VALUES ( Value1, Value2 ), ( Value1, Value2 )`
`INSERT INTO MyTable ( ID, Column1, Column2 ) VALUES ( 71, Value1, Value2 ), ( 78, Value1, Value2 )`
The next one, is without any restriction, just adding the Identity column. 
I don't know how it's made by you guys, but seems to be easy(I know is not so)
Thank you @sdanyliv for the discussion
Thank you @MaceWindu for adding this to 2.0.0 version. Waiting for it ;)

**Comment by sdanyliv** [comment]
As i have mentioned before about providers evolution, for now Npgsql provider has ability to insert data faster
http://www.npgsql.org/doc/copy.html
If you have time you can propose PR with this implementation :)
`OracleBulkCopy` is good start to understand how to do that dynamically.

**Comment by YXMagnuMXY** [comment]
@MaceWindu What means if the issue is closed? Is just answered well or the fix will be included in the new update?

**Comment by MaceWindu** [comment]
yes, fix was just merged into master and you can try it from myget

**Comment by hassanselim0** [comment]
@MaceWindu Is it possible to port this fix to linq2db v1.10?

**Comment by MaceWindu** [comment]
We don't have plans for new 1.x releases. I would recommed to migrate to 2.0 prerelease version as we are not aware of any regresssions in it if you don't plan to use new CTE feature, which still needs polishing.
